### ISSUE TITLE
[JENKINS-69751] Priority is disregarded if two blocked jobs are added to the queue

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/priority-sorter-plugin#623]
<h2><a name="Problemdescriptionbyusinganexample%3A%C2%A0"></a>Problem description by using an example: </h2>

<p>Build graph: </p>

<pre>
1─┐
│ │
▼ │
2 │
│ │
▼ │
3◄┘ 
 </pre>
<p>1. The number represents the priority of the projects. <br/>
2. The flags "Block build when upstream project is building" and "Block build when downstream project is building" are activated for all three projects. </p>

<p>If I trigger project 1 I expect the following build order: <br/>
1 -> 2 -> 3</p>

<p>I actually get the following build order: <br/>
1 -> 3 -> 2 -> 3</p>

<h2><a name="Analysis%3A%C2%A0"></a>Analysis: </h2>

<p>After looking at the code of Jenkins and PrioritySorter, I think I figured out what happens: </p>

<p>After project 1 is done, both project 2 and project 3 enter the waitingList in the queue. The projects enter the waitingList based on their alphabetical order. In this example project 3 enters before project 2. Since the "block builds"-flags are activated, both projects are moved from the waitingList to the blockedProjects <span class="error">&#91;^1&#93;</span>. </p>

<p>The blockedProjects are now sorted by using FIFO. Since both projects are now blocked, the AbstractProject.getCauseOfBlockage() realizes that both projects are in a deadlock, and the first considered project (project 3 in this case), is being unblocked. The snapshot is immediately updated and thus the second project (project 2 in this case) remains on the blockedProject-Collection. </p>

<p>Since there is now only one project on the buildables collection, sorting by the PrioritySorter has no effect and thus the project with the lower priority is being built before the project with the higher priority. </p>

<p><span class="error">&#91;^1&#93;</span>: One might assume that only one project is moved to blockedProjects and the other is moved to buildables, but this isn't the case for the following reason: When Queue.maintain() checks if a project can be moved from the waitingList to buildable/blockedProjects, the snapshot of the queue has not yet been updated. Thus the first project is <em>both</em> in the waitingList and in the blockedProjects when the second project is checked. </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=simon_sohrt">simon_sohrt</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-69751" target="_blank">Priority is disregarded if two blocked jobs are added to the queue</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Minor
<li><b>component(s)</b>: prioritysorter-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 1
<li><b>watchers</b>: 2
<li><b>imported</b>: 20251211-071809
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<h2><a name="Problemdescriptionbyusinganexample%3A%C2%A0"></a>Problem description by using an example: </h2>

<p>Build graph: </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">1─┐
│ │
▼ │
2 │
│ │
▼ │
3◄┘ 
 </pre>
</div></div>
<p>1. The number represents the priority of the projects. 
2. The flags "Block build when upstream project is building" and "Block build when downstream project is building" are activated for all three projects. </p>

<p>If I trigger project 1 I expect the following build order: 
1 -&gt; 2 -&gt; 3</p>

<p>I actually get the following build order: 
1 -&gt; 3 -&gt; 2 -&gt; 3</p>

<h2><a name="Analysis%3A%C2%A0"></a>Analysis: </h2>

<p>After looking at the code of Jenkins and PrioritySorter, I think I figured out what happens: </p>

<p>After project 1 is done, both project 2 and project 3 enter the waitingList in the queue. The projects enter the waitingList based on their alphabetical order. In this example project 3 enters before project 2. Since the "block builds"-flags are activated, both projects are moved from the waitingList to the blockedProjects <span class="error">&#91;^1&#93;</span>. </p>

<p>The blockedProjects are now sorted by using FIFO. Since both projects are now blocked, the AbstractProject.getCauseOfBlockage() realizes that both projects are in a deadlock, and the first considered project (project 3 in this case), is being unblocked. The snapshot is immediately updated and thus the second project (project 2 in this case) remains on the blockedProject-Collection. </p>

<p>Since there is now only one project on the buildables collection, sorting by the PrioritySorter has no effect and thus the project with the lower priority is being built before the project with the higher priority. </p>

<p><span class="error">&#91;^1&#93;</span>: One might assume that only one project is moved to blockedProjects and the other is moved to buildables, but this isn't the case for the following reason: When Queue.maintain() checks if a project can be moved from the waitingList to buildable/blockedProjects, the snapshot of the queue has not yet been updated. Thus the first project is <em>both</em> in the waitingList and in the blockedProjects when the second project is checked. </p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins 2.367 <br/>
Priority Sorter 4.1.0
```
</details>
